### ISSUE TITLE
Dashboard chat repeatedly shows "gateway exited, recovering your session" — ui-tui processes SIGHUP-killed and orphaned

### ISSUE THREAD

**Issue by Memofus** [NousResearch/hermes-agent#110164]
**Version:** 0.21.2
**Deployment:** Docker (s6-supervised), dashboard bound `0.0.0.0:9119` behind an nginx/FRP reverse proxy
**Surface affected:** Web dashboard chat (and desktop/TUI chat). Weixin/CLI/API are unaffected.

## Summary

During normal dashboard chat, the client frequently shows:

> gateway exited, recovering your session (any in-flight reply was lost)

…but the session is **not** actually gone: opening a new conversation and entering the session via `/session` shows the turn is still working. The *main gateway* (`hermes gateway run`) stays up the whole time — only the per-session TUI backend child dies.

## Root cause (as far as traced)

Dashboard chat spawns a per-session `ui-tui` Node.js process (`/opt/hermes/ui-tui/dist/entry.js`, child of `hermes dashboard`), which in turn manages a Python `tui_gateway` child. These processes are being repeatedly killed by **SIGHUP** and then respawned, leaving orphaned processes behind:

- `ui-tui/src/gatewayClient.ts` logs `graceful-exit received signal=SIGHUP → killing gateway` and calls `GatewayClient.kill`.
- The PTY/terminal lifecycle (`hermes_cli/pty_session.py`: `run_reaper` every 60s, `ttl=30*60`, `reap_idle`) closes the pty, which makes the kernel send SIGHUP to the foreground process group on that pty (the `ui-tui` process).
- The `tui_gateway` child (`tui_gateway/entry.py`) also reads stdin in a loop; `_stdin_recovery.py` documents a spurious-EOF race where a tool-call subprocess flips `O_NONBLOCK` on the shared fd 0, causing an apparent EOF and process exit (capped at `MAX_RECOVERIES_PER_MINUTE = 10`).

The client then shows the dropped WebSocket (code 1006) as "gateway exited".

## Evidence

- `logs/tui_gateway_crash.log`: 41+ occurrences of
  `[tui-parent] … graceful-exit received signal=SIGHUP → killing gateway`.
- `logs/gui.log`: repeated `ws closed … code=1006`, `reaped_sessions=1`, `detached_sessions=1`, `pty accepted/closed`.
- `logs/errors.log`: `WebSocketDisconnect(code=1006)` raised in `hermes_cli/web_routers/chat_ws.py:_broadcast_event` (`await sub.send_text(payload)`).
- Orphan accumulation: at one point **13** `ui-tui/dist/entry.js` processes were running (each ~180 MB RSS ≈ 2.3 GB total), all with `ppid=<dashboard pid>` and dead gateway children (`GatewayClient.kill … pid=none`).
- The orphan count kept climbing (~1 new process every 1–2 min) while a dashboard tab was connected.

## What was tried (no effect)

- Raised `proxy_read_timeout` to 300s on **both** reverse-proxy hops (public VPS nginx and NAS nginx). No change — because this is not a network idle timeout; it is server-side process lifecycle (SIGHUP on pty close).

## Expected vs actual

- **Expected:** A long-running or multi-turn dashboard chat keeps a single stable `ui-tui`/`tui_gateway` session until the user closes it.
- **Actual:** Sessions are reaped/SIGHUP-killed mid-turn (drops correlate with long turns of ~2–6 min), the client misreports "gateway exited", and orphaned `ui-tui` processes accumulate instead of being cleaned up.

## Temporary workarounds

- Use a non-dashboard surface (Weixin, CLI, or API) — these do not go through `ui-tui`.
- Kill the orphaned `ui-tui` processes to reclaim memory:
  `pkill -f '/opt/hermes/ui-tui/dist/entry.js'` (note: `-f` also matches the invoking shell, so use `pgrep … | xargs kill` or target pids explicitly).

## Notes for triage

- `hermes_cli/web_server.py` disables the WS keepalive ping on loopback binds (`_is_loopback`), but this dashboard binds `0.0.0.0`, so ping is enabled — the drops are therefore not explained by ping starvation here; the SIGHUP/pty-reap path is the dominant cause observed.
- The two candidate mechanisms (pty SIGHUP reap vs. `tui_gateway` stdin-EOF race) may both contribute; the crash log shows SIGHUP as the primary signal, with stdin-EOF recovery machinery present but capped.

**Comment by kokhlo** [comment]
Verified against today's `main` (b9271bcb34) — the wiring in the report maps cleanly onto current code, and #110168 addresses the right surface. Adding a few findings from reading main that don't fit the report's causal order, since they affect how much of the symptom the PR closes.

**1. The pty reaper cannot reap a mid-turn (attached) session on current main.**
`PtySessionRegistry.reap_idle` (`hermes_cli/pty_session.py:187-199`) only closes sessions where `not s.alive` (child already exited) or `not attached and last_detached_at + ttl` — with `ttl=30*60` (`hermes_cli/web_server_chat.py:48-49`), and the clock only starts after the WS detaches (`chat_ws.py:549-552` detaches, never closes). A tab that's connected is `attached=True` and is skipped on every 60s reaper tick. So "the reaper closes the pty → kernel SIGHUPs ui-tui **mid-turn**" is not reachable while the tab is attached — yet the reported drops correlate with live turns of 2–6 min, far inside the 30-min TTL. If the crash log really shows `graceful-exit signal=SIGHUP` while the tab was still connected, the first domino is upstream of the registry: either the WS dropped first (`code=1006`, detach, and then something closes the pty), or the SIGHUP never came from the pty at all — in an s6-supervised container the supervision tree itself is a candidate signal source that neither the issue nor the PR lists.

**2. The orphan count is accounted for by the TTL, not by a leak in `GatewayClient.kill`.**
13 orphans at ~1 new process per 1–2 min ≈ the 30-minute `ttl`: each dropped tab leaves a detached-but-alive PTY session whose ui-tui sits in the inert "gateway exited" state — recovery is bounded (`planGatewayRecovery`, 3 attempts / 60s, `ui-tui/src/app/gatewayRecovery.ts`) so there's no spawn storm; the Node process simply stays alive on its stdin listener (the same reason `die()` needs an explicit `process.exit(0)`, `useMainApp.ts:528-538`). The reaper eventually signals them when the TTL lapses — which is consistent with SIGHUP being the *dominant* signal in the crash log and with `pid=none` in the kill lines.

**3. For #110168's approach — one property worth an explicit test.**
Ignoring `SIGHUP` in `DASHBOARD_TUI_MODE` (symmetric to the existing `SIGINT` ignore at `entry.tsx:106-110`) is reasonable *because* the pty death has an independent exit guarantee: once the master side is really gone, every stdout/stderr write turns EIO/EPIPE and the `onError` bail-out in `entry.tsx` (`dead output stream ... → exiting`, 5 consecutive) reaps the process. That's the safety net that keeps a SIGHUP-ignoring embedded TUI from becoming a zombie — worth asserting in the PR so a future refactor of `onError` can't silently remove it.

Two checks that would pin down the remaining unknown (the true SIGHUP source) regardless of the PR:
- In `tui_gateway_crash.log`, do the `SIGHUP` lines for a given session **precede or follow** the corresponding `ws closed code=1006` in `gui.log`? Before → server-side origin (pty/s6); after → WS-drop cascade.
- Does the cadence change with the dashboard reached directly (SSH tunnel to :9119, both proxies bypassed) for an hour of long turns?

Happy to review #110168 — from reading main it composes with both candidate sources, it just shouldn't be read as closing the "who sends the SIGHUP" question.

**Comment by Memofus** [comment]
Ran the first check — timeline correlation across `tui_gateway_crash.log` and `gui.log`.

**The `transport exit code=1006` (crash.log) and `ws closed code=1006` (gui.log) are the same event, seen from both sides.** They align within milliseconds:

| crash.log (`[tui-parent]`) | gui.log (`tui_gateway.ws`) | delta |
|---|---|---|
| 17:00:36.674 transport-exit-1006 | 17:00:36.671 ws closed 1006 | 3 ms |
| 17:03:34.348 transport-exit-1006 | 17:03:34.345 ws closed 1006 | 3 ms |
| 17:26:21.117 transport-exit-1006 | 17:26:21.341 ws closed 1006 | 224 ms |

These are the loopback (`127.0.0.1`) WS between the `ui-tui` process and the gateway.

**The SIGHUP lines consistently come *after* the 1006 drop, not before:**

| 1006 drop | subsequent SIGHUP | delta |
|---|---|---|
| 17:26:21 | 17:27:42 | +81 s |
| 17:35:02 | 17:35:21 | +19 s |
| 17:37:25 | 17:37:43 | +18 s |

And the SIGHUPs often land on a ~60 s cadence (17:31:42 / 17:33:43 / 17:35:21 / 17:36:43 / 17:37:43), matching `run_reaper`'s 60 s interval.

So the causal order is: **loopback WS drops with 1006 first → session detaches (`detached_sessions=1`) → the 60 s reaper later SIGHUPs the orphaned `ui-tui`**. The pty/SIGHUP is downstream cleanup, not the first domino — consistent with your point 1.

That narrows the open question to: **what terminates the loopback (`127.0.0.1`) WS with 1006 mid-turn?** The reaper can't be it (attached sessions are skipped), and the loopback path has no proxy/network in between. Will run your second check (direct SSH tunnel to :9119, both proxies bypassed, ~1 h of long turns) when I can.

**Comment by 3oltan-seo** [comment]
Independent entry into the same "attached forever" state, plus some numbers that may bear on the open question in the thread.

**1. A client that drops during the attach replay leaves the session attached forever — filed separately as #110849.**
`PtySession.attach()` sets `self._ws` / `attached = True` / `last_detached_at = None` *before* `await ws.send_bytes(snap)`, and `pty_ws` calls it outside the `try` that owns `PTY_REGISTRY.detach()`. Trace from a dashboard server log for the client-side `/api/pty` socket:

```
chat_ws.py:516 in pty_ws
pty_session.py:111 in attach
    await ws.send_bytes(snap)
starlette/websockets.py:89 in send -> raise WebSocketDisconnect(code=1006)
...
uvicorn.protocols.utils.ClientDisconnected
```

The session is then left at `attached=True, last_detached_at=None` — the one combination `reap_idle()` can never reclaim (`not attached` is required for the TTL branch, and the TTL clock is anchored to `last_detached_at`, which stays `None`). No SIGHUP cascade, no proxy and no ping setting involved. Note this is the client-side `/api/pty` socket, *not* the `ui-tui ↔ gateway` loopback WS that drops first in your timeline tables — so it can add orphans on top of that sequence rather than explaining it.

**2. Quantifying the `GatewayClient.kill … pid=none` line from the report.**
Across `tui_gateway_crash.log`: 635 `GatewayClient.kill` lines, **587** of them `pid=none killResult=none` (92%). That is expected by construction — `describeChild(null)` prints `pid=none` (`ui-tui/src/gatewayClient.ts:49-52`) and `start()` nulls `this.proc` (`:614`) once the previous child was replaced or errored, so the graceful-exit path reports killing a gateway it no longer tracks and a still-live child survives its parent. A fix would keep pid + start time and signal the whole process group (SIGHUP → SIGTERM → SIGKILL), the way `hermes_cli/pty_bridge.py` already does in `PtyBridge.close()`; the repo already uses that ownership pattern for MCP children in `tools/mcp_death_supervisor.py`.

**3. Both halves of the pair leak independently.**
At peak we measured 18 `ui-tui` nodes / 15 `tui_gateway` children under one `hermes dashboard` with a single real terminal (`who` → console + one SSH). In both that snapshot and a rescan 45 minutes later, 2–3 of the nodes had **no** python child at all — one of them dated five days back. So the parent outliving its gateway (point 2) and the gateway outliving its parent are both observable, not just one direction.

Fix for the attach path is small (issue #110849): commit the attached state only after the replay write succeeds (or roll back in `except`), wrap the `session.attach()` call site in `try/except` with a detach, and/or give `reap_idle()` a criterion that does not depend on `attached`.
